### PR TITLE
Generalize shift below merge to multi left shift

### LINKED ISSUE #2133: QScript needs to merge `LeftShift` containing `Typecheck`
In the query
```
select * from nested_foo where (foo{*} = 5 or foo[*] = 6)
```
we need to merge the two sides of the `or` in qscript. Currently this fails to merge in qscript and creates an unnecessary theta join.

Here is the relevance `LogicalPlan`:
```
      ╰─ Invoke(Or)
         ├─ Invoke(Eq)
         │  ├─ Invoke(FlattenMap)
         │  │  ╰─ Let('__tmp2)
         │  │     ├─ ObjectProject(Free('__tmp0){"foo"})
         │  │     ╰─ Typecheck(Obj(Map[], Some(Top)))
         │  │        ├─ Free('__tmp2)
         │  │        ├─ Free('__tmp2)
         │  │        ╰─ Constant(Obj(Map( -> NA)))
         │  ╰─ Constant(Int(5))
         ╰─ Invoke(Eq)
            ├─ Invoke(FlattenArray)
            │  ╰─ Let('__tmp3)
            │     ├─ ObjectProject(Free('__tmp0){"foo"})
            │     ╰─ Typecheck(FlexArr(0, None, Top))
            │        ├─ Free('__tmp3)
            │        ├─ Free('__tmp3)
            │        ╰─ Constant(Arr(List(NA)))
            ╰─ Constant(Int(6))
```
This should enable the integration test `multitypeFlatten.test`. It is also possible that this work will enable a handful of other flatten-related integration tests.

### LINKED ISSUE #3092: Adjacent LeftShift collapse
Subsumes part of #2998 

Look for a join on two `LeftShift` nodes with a common source.  Put them together one-above-the-other (which one doesn't matter), ensuring that original value access works.  This will fix the issue in #2935 and likely also #2366

Obviously this will be implemented as a rewrite on the graph, probably after `ReifyProvenance`.

### LINKED ISSUE #3222: Avoid thetajoins in cases with differing dimensions
Easy example:

```sql
select a[*] + b from foo
```

This really shouldn't be a thetajoin.  We can rearrange the `ProjectKey` with `b` to happen after the shift on the `LeftSide` of the results.  This shows up in other areas as well.  For example:

```sql
select a[*][*] + b[*] + c[*][*][*] from foo
```

Same case, just bigger.  This is a generalization of autojoin elimination which often gets confused with #3092.  Resolving this issue is important for the search card.

### PR DESCRIPTION
**Pull request by djspiewak**
I'm just going to leave this right here…

```sql
select a[*][*][*] + b - c[*] / d[*][*] from foo
```

QScript: https://gist.github.com/7024e91afdcaeee9b016ef427d5d5117

<kbd>Cmd</kbd>-<kbd>F</kbd> `theta`

*No Results Found*

:tada:

Ping @jdegoes @alissapajer 

Depends on #3299
Fixes #2133
Fixes #3092
Fixes #3222

### CODE REVIEW

**Comment by djspiewak** [comment]
Ok this is an extremely bizarre commit list.  Also I was excited and filed my PR and didn't wait for @edmundnoble's conflict resolution.  So uh… don't merge yet.

**Comment by djspiewak** [comment]
I think this is failing because of the `ShiftArray` issue.

**Comment by jdegoes** [comment]
> select a[*][*][*] + b - c[*] / d[*][*] from foo

Sweeeet! 🎉

**Comment by djspiewak** [comment]
@wemrysi If you can do a final review and merge on this tonight, that would be *greatly* appreciated.  I'd like to get it into master asap so we can get the build cycles through and have time tomorrow morning to run the downstream builds, which also take a fair amount of time.

File: connector/src/main/scala/quasar/qscript/qsu/minimizers/CollapseShifts.scala [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+          case (-\/(ls), idx) => ls.repair.map {
+            case QSU.LeftTarget() => scala.sys.error("QSU.LeftTarget in CollapseShifts")
+            case QSU.AccessLeftTarget(access) => access.left[Int]
+            case QSU.RightTarget() => idx.right[QAccess[Hole]]
+          }
+          case (\/-(mls), idx) => mls.repair.map(_.map(_ + idx))
+        }
+        val shiftsLength = shifts.length
+        val repair = fm.flatMap { idx =>
+          // get is safe because fm's indices line up with updatedShifts'
```
Comment (wemrysi) [inline connector/src/main/scala/quasar/qscript/qsu/minimizers/CollapseShifts.scala]: Why not use it then?

File: connector/src/test/scala/quasar/qscript/qsu/ExpandShiftsSpec.scala [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+            func.Eq(
+              func.AccessLeftTarget(Access.id(IdAccess.identity[Fix[EJson]]('qsu0), _)),
+              func.AccessLeftTarget(Access.id(IdAccess.identity[Fix[EJson]]('qsu1), _))),
+            func.ConcatMaps(
+              func.AccessLeftTarget(Access.valueHole(_)),
+              func.MakeMapS("1", func.RightTarget)),
+            func.Undefined
+          )
+        )
+        ok
```
Comment (wemrysi) [inline connector/src/test/scala/quasar/qscript/qsu/ExpandShiftsSpec.scala]: Can this (and all the others) be removed?

File: connector/src/main/scala/quasar/qscript/qsu/minimizers/CollapseShifts.scala (hunk 1 above)
Comment (djspiewak) [reply connector/src/main/scala/quasar/qscript/qsu/minimizers/CollapseShifts.scala]: It requires enabling more warts.

**Comment by djspiewak** [comment]
The other comments are basically re @edmundnoble. Flagging for later attention

File: connector/src/main/scala/quasar/qscript/qsu/minimizers/CollapseShifts.scala (hunk 1 above)
Comment (wemrysi) [reply connector/src/main/scala/quasar/qscript/qsu/minimizers/CollapseShifts.scala]: I'd consider that a feature 😄

**Pull request merged**